handle free variables correctly

hiratara opened the pull request:
Our current implementation of Term doesn't have free variable names as it is only a De Bruijn index expression, which is not correct. Free variable names are critical for open terms.

The issue is demonstrated by the following test, where free variable names were lost completely:

https://github.com/ljedrz/lambda_calculus/blob/c961a869398c1710dde5a609ee41a94d3f71cd74/tests/reduction.rs#L62

This PR introduces a naming Context to give names to free variables, which is the same approach as the TAPL book.

## Notes

A key inspiration for this change is the original TAPL implementation, which uses a top-level syntax to define variables (see https://github.com/mspertus/TAPL/blob/main/untyped/test.f ). In that implementation, you first have to define a variable like `x` (using `x/;` ), which pushes `x` into the naming context. Then, you can use `x` as a free variable.

Since our implementation does not have such a top-level syntax, this PR introduces a Context type instead, which must be prepared manually by the user.

hiratara committed:
feat: add `Context` for handling free variables

* Introduce `Context` to define free variables for parsing and display
* Add `parse_with_context` to to resolve free variables based on a `Context`
* Add `Term::with_context` for displaying terms using a given `Context`

hiratara committed:
refactor: Remove unnecessary use of VecDeque

revert 190c29f182c0d6e66a05a8c00bf23f46fd27bc66

hiratara committed:
test: add unit tests for Context and related features

ljedrz reviewed (commented):
Yeah, I think this feature is long overdue in this repo; would you like to fix the `clippy` warnings and consider one suggestion I've left?

Thanks!

hiratara committed:
refactor: use slice for function parameter

Accepting a slice is more flexible and idiomatic than a vector reference (@ljedrz)

Co-authored-by: ljedrz <3750347+ljedrz@users.noreply.github.com>

hiratara committed:
fix: address clippy::doc_lazy_continuation warning in doc comments

hiratara committed:
fix: make lifetime syntax consistent per Clippy

hiratara committed:
fix: remove redundant tests causing clippy warnings

ljedrz reviewed (approved):
Looks good, thanks again!
